### PR TITLE
session: read local dc replicas automatically for stale read

### PR DESCRIPTION
**Pull request by Yisaer**
### What is changed and how it works?

In previous, we should use `txn_scope` to enable read local dc replicas for stale read which may expose the local transaction feature while it is not GA yet.

This request also updated tikcClient, see https://github.com/tikv/client-go/pull/89

What's Changed:

Read local dc replicas automatically for stale read by detecting the `zone` label configuration for the tidb-server.

Also the `select as of timestamp` statement is not supported local stale read, this request make it supported.

### Check List 

Tests 

- Unit test

### Release note 

- No release note

### CODE REVIEW

File: executor/builder.go [hunk 1]
```diff
@@ -95,14 +96,17 @@ type CTEStorages struct {
...
-	return &executorBuilder{
+func newExecutorBuilder(ctx sessionctx.Context, is infoschema.InfoSchema, ti *TelemetryInfo, snapshotTS uint64, explicitStaleness bool, txnScope string) *executorBuilder {
+	e := &executorBuilder{
 		ctx:               ctx,
 		is:                is,
 		Ti:                ti,
 		snapshotTS:        snapshotTS,
 		explicitStaleness: explicitStaleness,
 	}
+	e.txnScope = txnScope
```
Comment (xhebox) [inline executor/builder.go:107]: Just return? There is no need for new variable.

File: planner/core/preprocess.go [hunk 2]
```diff
@@ -1456,9 +1459,31 @@ func (p *preprocessor) checkFuncCastExpr(node *ast.FuncCastExpr) {
...
-// If as of is not nil, timestamp is used to get the history infoschema from the infocache.
 func (p *preprocessor) handleAsOfAndReadTS(node *ast.AsOfClause) {
+	// When statement is during the Txn, we check whether there exists AsOfClause. If exists, we will return error,
+	// otherwise we should directly set the return param from TxnCtx.
+	if p.ctx.GetSessionVars().InTxn() {
+		if node != nil {
+			p.err = ErrAsOf.FastGenWithCause("as of timestamp can't be set in transaction.")
+			return
+		}
+		if p.initedLastSnapshotTS {
```
Comment (xhebox) [inline planner/core/preprocess.go:1470]: This line is wrong. This check is not only for init, but also check for different tso. It should not early return.

Comment (Yisaer) [reply planner/core/preprocess.go]: updated.

File: kv/txn_scope_var.go [hunk 3]
```diff
@@ -52,6 +52,9 @@ func (t TxnScopeVar) GetVarValue() string {
 
 // GetTxnScope returns the value of the tidb-server holds to request tso to pd.
 func (t TxnScopeVar) GetTxnScope() string {
+	if t.varValue == GlobalTxnScope {
```
Comment (nolouch) [inline kv/txn_scope_var.go:55]: why need this?  add a comment.

File: planner/core/preprocess.go [hunk 4]
```diff
@@ -1456,9 +1459,28 @@ func (p *preprocessor) checkFuncCastExpr(node *ast.FuncCastExpr) {
...
+	if p.ctx.GetSessionVars().InTxn() {
+		if node != nil {
+			p.err = ErrAsOf.FastGenWithCause("as of timestamp can't be set in transaction.")
+			return
+		}
+		txnCtx := p.ctx.GetSessionVars().TxnCtx
+		p.SnapshotTSEvaluator = func(sessionctx.Context) (uint64, error) {
+			return txnCtx.StartTS, nil
+		}
+		p.LastSnapshotTS = txnCtx.StartTS
```
Comment (nolouch) [inline planner/core/preprocess.go:1471]: Why need to set it, I think it breaks the `getSnapshotTS()`'s logic.

File: executor/table_reader.go [hunk 5]
```diff
@@ -324,6 +326,7 @@ func (e *TableReaderExecutor) buildKVReqSeparately(ctx context.Context, ranges [
 			SetDesc(e.desc).
 			SetKeepOrder(e.keepOrder).
 			SetStreaming(e.streaming).
+			SetTxnScope(e.txnScope).
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply executor/table_reader.go:329]: This build requests for tiflash.

File: infoschema/cache.go [hunk 6]
```diff
@@ -52,12 +62,12 @@ func (h *InfoCache) GetLatest() InfoSchema {
 func (h *InfoCache) GetByVersion(version int64) InfoSchema {
 	h.mu.RLock()
 	defer h.mu.RUnlock()
-	metrics.InfoCacheCounters.WithLabelValues("get", "version").Inc()
+	getVersionCounter.Inc()
```
Comment (nolouch) [inline infoschema/cache.go:65]: This is from other PRs.

File: planner/core/preprocess.go (hunk 4 above)
Comment (xhebox) [reply planner/core/preprocess.go:1471]: Kind of. He wants to specify the ts clearly from the preprocessor down to executor. Need a `IsStaleness` guard to ensure the logic only applies to stale txn.

**Review by xhebox — COMMENTED** [review]
Rest LGTM

File: planner/core/preprocess.go (hunk 4 above)
Comment (Yisaer) [reply planner/core/preprocess.go:1471]: updated.

**Review by xhebox — APPROVED** [review]

**Review by nolouch — APPROVED** [review]
lgtm

**Pull request merged**